### PR TITLE
[Merged by Bors] - Eisenstein series are  mdifferentiable

### PR DESCRIPTION
**Pull request by CBirkbeck**
We show that Eisenstein Series are MDifferentiable

- [x] depends on: #10377 
- [x] depends on:  #11244

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by CBirkbeck** [comment]
I'm not sure if there is a more general version of `MDifferentiable_iff_extension_DifferentiableOn` that we might want to add. I experimented with this, but couldn't quite get the setup right.

File: Mathlib/Analysis/Complex/UpperHalfPlane/Manifold.lean [hunk 1]
```diff
@@ -34,4 +35,40 @@ theorem mdifferentiable_coe : MDifferentiable 𝓘(ℂ) 𝓘(ℂ) ((↑) : ℍ 
...
+          (OpenEmbedding.toPartialHomeomorph UpperHalfPlane.coe openEmbedding_coe))
+
+/--This shows that being MDifferentiable as a map `ℍ → ℂ` is equivalent to being
+differentiable on `{z : ℂ | 0 < z.im}` after arbitrarily extending to a function on all of `ℂ`.-/
+lemma MDifferentiable_iff_extension_DifferentiableOn (f : ℍ → ℂ) : MDifferentiable 𝓘(ℂ) 𝓘(ℂ) f ↔
+    DifferentiableOn ℂ (↑ₕf) (UpperHalfPlane.coe '' ⊤) := by
+  rw [_root_.MDifferentiable]
+  constructor
+  · intro h z hz
+    obtain ⟨y, _, hy⟩ := hz
```
Comment (riccardobrasca) [inline Mathlib/Analysis/Complex/UpperHalfPlane/Manifold.lean:48]:
```suggestion
  · intro h z ⟨y, _, hy⟩
```

File: Mathlib/Analysis/Complex/UpperHalfPlane/Manifold.lean [hunk 2]
```diff
@@ -34,4 +35,40 @@ theorem mdifferentiable_coe : MDifferentiable 𝓘(ℂ) 𝓘(ℂ) ((↑) : ℍ 
...
+    simp only [top_eq_univ, mem_univ, ← hy, writtenInExtChartAt, extChartAt, extend,
+      refl_partialEquiv, PartialEquiv.refl_source, singletonChartedSpace_chartAt_eq,
+      modelWithCornersSelf_partialEquiv, PartialEquiv.trans_refl, PartialEquiv.refl_coe,
+      OpenEmbedding.toPartialHomeomorph_source, coe_coe_symm, CompTriple.comp_eq,
+      modelWithCornersSelf_coe, range_id, toFun_eq_coe, OpenEmbedding.toPartialHomeomorph_apply,
+      image_univ] at H ⊢
+    apply H.2.mono (Set.subset_univ _)
+  · intro h z
+    have ha : UpperHalfPlane.coe '' ⊤ ∈ 𝓝 ↑z := by
+      exact IsOpenMap.image_mem_nhds (OpenEmbedding.isOpenMap openEmbedding_coe) (by simp)
```
Comment (riccardobrasca) [inline Mathlib/Analysis/Complex/UpperHalfPlane/Manifold.lean:60]:
```suggestion
    have ha : UpperHalfPlane.coe '' ⊤ ∈ 𝓝 ↑z :=
      IsOpenMap.image_mem_nhds (OpenEmbedding.isOpenMap openEmbedding_coe) (by simp)
```

File: Mathlib/NumberTheory/ModularForms/EisensteinSeries/MDifferentiable.lean [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+lemma div_linear_zpow_differentiableOn (k : ℤ) (a : Fin 2 → ℤ) :
+    DifferentiableOn ℂ (fun z : ℂ => 1 / (a 0 * z + a 1) ^ k) {z : ℂ | 0 < z.im} := by
+  by_cases ha : a ≠ 0
+  · apply DifferentiableOn.div (differentiableOn_const 1)
+    · apply DifferentiableOn.zpow
+      fun_prop
+      left
+      exact fun z hz ↦ linear_ne_zero ((Int.cast (R := ℝ)) ∘ a) ⟨z, hz⟩
+        ((Function.comp_ne_zero_iff _ Int.cast_injective Int.cast_zero ).mpr ha)
+    · exact fun z hz ↦ zpow_ne_zero k (linear_ne_zero ((Int.cast (R := ℝ)) ∘ a)
```
Comment (loefflerd) [inline Mathlib/NumberTheory/ModularForms/EisensteinSeries/MDifferentiable.lean:43]:
```suggestion
    · exact fun z hz ↦ zpow_ne_zero k (linear_ne_zero (a ·)
```

File: Mathlib/Analysis/Complex/UpperHalfPlane/Manifold.lean [hunk 4]
```diff
@@ -34,4 +35,39 @@ theorem mdifferentiable_coe : MDifferentiable 𝓘(ℂ) 𝓘(ℂ) ((↑) : ℍ 
   smooth_coe.mdifferentiable
 #align upper_half_plane.mdifferentiable_coe UpperHalfPlane.mdifferentiable_coe
 
+local notation "↑ₕ" f => f ∘ (PartialHomeomorph.symm
+          (OpenEmbedding.toPartialHomeomorph UpperHalfPlane.coe openEmbedding_coe))
+
+/--This shows that being MDifferentiable as a map `ℍ → ℂ` is equivalent to being
+differentiable on `{z : ℂ | 0 < z.im}` after arbitrarily extending to a function on all of `ℂ`.-/
+lemma MDifferentiable_iff_extension_DifferentiableOn (f : ℍ → ℂ) : MDifferentiable 𝓘(ℂ) 𝓘(ℂ) f ↔
+    DifferentiableOn ℂ (↑ₕf) (UpperHalfPlane.coe '' ⊤) := by
```
Comment (loefflerd) [inline Mathlib/Analysis/Complex/UpperHalfPlane/Manifold.lean:44]:
```suggestion
    DifferentiableOn ℂ (↑ₕf) {z : ℂ | 0 < z.im} := by
```

File: Mathlib/Analysis/Complex/UpperHalfPlane/Manifold.lean [hunk 5]
```diff
@@ -34,4 +35,39 @@ theorem mdifferentiable_coe : MDifferentiable 𝓘(ℂ) 𝓘(ℂ) ((↑) : ℍ 
   smooth_coe.mdifferentiable
 #align upper_half_plane.mdifferentiable_coe UpperHalfPlane.mdifferentiable_coe
 
+local notation "↑ₕ" f => f ∘ (PartialHomeomorph.symm
+          (OpenEmbedding.toPartialHomeomorph UpperHalfPlane.coe openEmbedding_coe))
+
+/--This shows that being MDifferentiable as a map `ℍ → ℂ` is equivalent to being
+differentiable on `{z : ℂ | 0 < z.im}` after arbitrarily extending to a function on all of `ℂ`.-/
```
Comment (loefflerd) [inline Mathlib/Analysis/Complex/UpperHalfPlane/Manifold.lean:42]: I'm surprised this is so painful. The comparable argument for Jacobi theta functions (in JacobiTheta.Manifold) is literally a one-liner, why doesn't the same work here?

Comment (CBirkbeck) [reply Mathlib/Analysis/Complex/UpperHalfPlane/Manifold.lean:42]: Ah I didn't know about `DifferentiableAt.mdifferentiableAt` which is the key here, this way I don't need that other ugly lemma which I was wondering about. I'll incorporate this and try again.

File: Mathlib/NumberTheory/ModularForms/EisensteinSeries/MDifferentiable.lean [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+namespace EisensteinSeries
+
+/-- Extend a function on `ℍ` arbitrarily to a function on all of `ℂ`. -/
+local notation "↑ₕ" f => f ∘ (PartialHomeomorph.symm
+          (OpenEmbedding.toPartialHomeomorph UpperHalfPlane.coe openEmbedding_coe))
+
+/--Auxilary lemma showing that for any `k : ℤ` the function `z → 1/(c*z+d)^k` is
+differentiable on `{z : ℂ | 0 < z.im}`. -/
+lemma div_linear_zpow_differentiableOn (k : ℤ) (a : Fin 2 → ℤ) :
+    DifferentiableOn ℂ (fun z : ℂ => 1 / (a 0 * z + a 1) ^ k) {z : ℂ | 0 < z.im} := by
```
[the remark this answers is not in the document]
Comment (CBirkbeck) [reply Mathlib/NumberTheory/ModularForms/EisensteinSeries/MDifferentiable.lean:35]: I think I picked this up from the mathsport and just stuck with it.

File: Mathlib/Analysis/Complex/UpperHalfPlane/Basic.lean [hunk 7]
```diff
@@ -72,6 +72,10 @@ instance canLift : CanLift ℂ ℍ ((↑) : ℍ → ℂ) fun z => 0 < z.im :=
   Subtype.canLift fun (z : ℂ) => 0 < z.im
 #align upper_half_plane.can_lift UpperHalfPlane.canLift
 
+lemma coe_image_eq : UpperHalfPlane.coe '' ⊤ = {z : ℂ | 0 < z.im} := by
+  convert (Subtype.coe_image (s := (⊤ : Set ℍ)))
+  simp only [Set.top_eq_univ, Set.mem_univ, exists_prop, and_true]
+
```
[the remark this answers is not in the document]
Comment (CBirkbeck) [reply Mathlib/Analysis/Complex/UpperHalfPlane/Basic.lean]: Yes with the other changes I no longer need this lemma.

**Comment by CBirkbeck** [comment]
I ran `lake exe shake --update` but I don't know if I should have, I'll wait for an expert to say if this is ok.

**Comment by loefflerd** [comment]
No, that's the wrong thing: you should revert your changes to `noshake.json` and run `lake exe shake --fix` instead.

**Comment by CBirkbeck** [comment]
> No, that's the wrong thing: you should revert your changes to `noshake.json` and run `lake exe shake --fix` instead.

I did that originally, but the fix meant a file wouldn't compile, so I went for the update route.

**Comment by loefflerd** [comment]
That's weird. If `lake exe shake --fix` is giving wrong results, I suggest that you restore the branch to the state it was in before you started the tree-shaking, and then report on zulip that the script is giving misleading output on this branch.

**Comment by loefflerd** [comment]
I have done some digging and I know what's causing this. Shall I push my fix?

**Comment by CBirkbeck** [comment]
> I have done some digging and I know what's causing this. Shall I push my fix?

Sure it seems it doesn't like the file where the notation is defined. I guess one solution is to move the notation from ../Manifold to ../Topology?

**Comment by loefflerd** [comment]
Not quite. The issue is that `notation` declarations are visible to the parser, but invisible to `shake` (which only notices when the notation is _used_, not when it's defined). 

I suggest the following in `UpperHalfPlane.Topology`:
- add the `PartialHomeomorph` import
- add the following definition:
  ```lean
  /-- A continuous section `ℂ → ℍ` of the natural inclusion map, bundled as a `PartialHomeomorph`. -/
  def ofComplex : PartialHomeomorph ℂ ℍ := (openEmbedding_coe.toPartialHomeomorph _).symm
  
  /-- Extend a function on `ℍ` arbitrarily to a function on all of `ℂ`. -/
  scoped[UpperHalfPlane] notation "↑ₕ" f => f ∘ ofComplex
  ```
Then `shake` won't complain, because it sees the `def` and accepts that as a reason to import `PartialHomeomorph`. I think it's useful anyway to give the section of the inclusion map a name: then we can formulate a lemma that it's continuous or mdifferentiable, etc.

File: Mathlib/NumberTheory/ModularForms/EisensteinSeries/MDifferentiable.lean [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+  lift z to ℍ using hz
+  exact extends_def _ _
+
+/-- Eisenstein series are MDifferentiable (i.e. holomorphic functions from `ℍ → ℂ`). -/
+theorem eisensteinSeries_SIF_MDifferentiable {k : ℤ} {N : ℕ} (hk : 3 ≤ k) (a : Fin 2 → ZMod N) :
+    MDifferentiable 𝓘(ℂ) 𝓘(ℂ) (eisensteinSeries_SIF a k) := by
+  intro τ
+  suffices DifferentiableAt ℂ (↑ₕeisensteinSeries_SIF a k) τ.1 by
+    convert MDifferentiableAt.comp τ (DifferentiableAt.mdifferentiableAt this) τ.mdifferentiable_coe
+    exact funext fun z ↦ (extends_def (eisensteinSeries_SIF a k) z).symm
```
Comment (loefflerd) [inline Mathlib/NumberTheory/ModularForms/EisensteinSeries/MDifferentiable.lean:59]:
```suggestion
    exact funext fun z ↦ (comp_ofComplex (eisensteinSeries_SIF a k) z).symm
```

File: scripts/noshake.json [hunk 9]
```diff
@@ -182,10 +182,7 @@
   "Batteries.Util.LibraryNote",
   "Batteries.Util.ProofWanted",
   "Batteries.Util.TermUnsafe"],
- "ignoreAll":
- ["Batteries.Tactic.Basic",
-  "Mathlib.Mathport.Syntax",
-  "Mathlib.Tactic.Linter"],
+ "ignoreAll": ["Batteries.Tactic.Basic", "Mathlib.Mathport.Syntax"],
```
[the remark this answers is not in the document]
Comment (CBirkbeck) [reply scripts/noshake.json]: No this was just a glitch same below I think

File: Mathlib/Analysis/Complex/UpperHalfPlane/Manifold.lean [hunk 10]
```diff
@@ -14,6 +14,7 @@ import Mathlib.Geometry.Manifold.MFDeriv.Basic
 In this file we define the complex manifold structure on the upper half-plane.
 -/
 
+open Set Filter PartialHomeomorph
```
Comment (loefflerd) [inline Mathlib/Analysis/Complex/UpperHalfPlane/Manifold.lean:17]: a git glitch that slipped through the net

**Comment by loefflerd** [comment]
Looks good to me now. If you're happy with it, please label with "awaiting review" and "awaiting CI", and I'll flag it for merging if the CI run passes.

**Comment by CBirkbeck** [comment]
done!

**Comment by loefflerd** [comment]
maintainer merge

**Pull request closed without merging**